refactor: remove defclass DBWorker

RCmerci opened the pull request:
- use a simple `#js{"remoteInvoke" remote-function}` to replace the whole `defclass DBWorker`.
   `(Comlink/expose #js{"remoteInvoke" thread-api/remote-function})`
- use `def-thread-api` to define a API to expose to other threads (in this pr, expose APIs from db-worker to ui-thread)
- `(def-thread-api :api-ns/api-name ...)`. I added a clj-kondo hook to this keyword, so it supports jumping and querying references in the editor(with clojure-lsp)
- the transit-read/write for API return values has been unified, so many `ldb/read-transit-str` and `ldb/write-transit-str` have been removed

LATER:
- this pr only handles apis defined by db-worker, there're also apis defined by ui-thread(expose to db-worker).
   I will handle them in another pr

logseq-cldwalker commented:
> use def-thread-api to define a API to expose to other threads (in this pr, expose APIs from db-worker to ui-thread)
(def-thread-api :api-ns/api-name ...). I added a clj-kondo hook to this keyword, so it supports jumping and querying references in the editor(with clojure-lsp)

❤️

logseq-cldwalker commented on src/main/frontend/worker/db_worker.cljs:
```diff
@@ -442,494 +441,291 @@
+  (when-let [conn (worker-state/get-datascript-conn repo)]
+    (worker-export/get-all-pages repo @conn)))
+
+(def-thread-api :export/get-all-page->content
+  [repo]
+  (when-let [conn (worker-state/get-datascript-conn repo)]
+    (worker-export/get-all-page->content repo @conn)))
+
+(def-thread-api :undo-redo/undo
```
Would it be helpful long term to use standard namespace keywords rather than shortened namespace ones? We also have these in shortcut.config and they're harder to work with since they aren't really unique and are context dependent. Semantically they don't feel meaningful since namespaces get created arbitrarily with no specific criteria

RCmerci replied:
what about using namespace `thread-api` for all these keywords?

RCmerci replied:
update:
has been made to use the namespace "thread-api",
ensuring these keywords won't conflict with others serving different purposes.

logseq-cldwalker reviewed (commented):
Code changes look good. Encountered two bugs during QA:

* In db graphs, creating a page from cmd-k no longer navigates to newly created page
* Writing any content in file graphs isn't working:
<img width="305" alt="Screen Shot 2025-03-25 at 12 34 03 PM" src="https://github.com/user-attachments/assets/7b6fcc48-1048-40a4-8d9e-c29934b2c5bc" />

RCmerci commented:
@logseq-cldwalker I fixed 2 bugs you listed above
> In db graphs, creating a page from cmd-k no longer navigates to newly created page

this bug isn't limited to create-page alone.  
since ui-outliner-tx/transact! now returns value that no longer require `ldb/read-transit-str`, I've reviewed all instances of ldb/read-transit-str, removing unnecessary ones. Additionally, I've added a check within ldb/read-transit-str to log/error and return directly if the parameter isn't a transit-string

> Writing any content in file graphs isn't working

this [commit](https://github.com/logseq/logseq/pull/11790/commits/4dad310ec226ea658c281db0a21b9ff75a15040c) fixed it

RCmerci commented on src/main/frontend/db/async.cljs:
```diff
@@ -147,13 +147,12 @@
                name-or-uuid)]
     (if (or (:block.temp/fully-loaded? e) async-requested?)
       e
-      (when-let [^Object sqlite @db-browser/*worker]
```
updated:
I removed ~40 instances of code like `(when-let [worker @*db-worker] (worker ...))`.
use `<invoke-db-worker` to replace them.
> Is this actually necessary in most places or is it only one or two places that needed it before worker was defined

yes, I think so.
Additionally, I believe we haven't properly handled the failure case of (when-let ...),
only skipping certain logic execution.  
Therefore, I think when encountering `(nil? @*db-worker)`, we should fail fast and throw an error(because all the usages should happen after *db-worker inited).

logseq-cldwalker commented on deps/db/src/logseq/db/sqlite/util.cljs:
```diff
@@ -50,7 +50,11 @@
         reader (transit/reader :json {:handlers read-handlers*})]
-    (fn read-transit-str* [s] (transit/read reader s))))
+    (fn read-transit-str* [s]
+      (if (and (string? s) (identical? "[" (first s)))
```
A comment for this check could be helpful. Unclear why we pass invalid calls through silently rather than fail explicitly

RCmerci replied:
The reason I didn't throw an error here is:  
If I accidentally missed deleting an ldb/read-transit-str in this PR, we can still see the :invalid-transit-string message in the console,  
without affecting the original normal logic.  

another reason for not using `assert` here:  
If assert is used here but ldb/read-transit-str is used within `p/let` (without explicit `p/catch`), I found no error log appears in the console, which provides no help in identifying the issue.

logseq-cldwalker replied:
Gotcha. Good to know it was useful for refactoring. Hopefully we can delete it later when there aren't any more missed deletions

RCmerci replied:
yes, we can keep it for maybe 1 month. let me add a todo in comment

logseq-cldwalker reviewed (approved):
@RCmerci Thanks for this hefty refactor! 👍 💪 Confirmed 2 bugs are fixed. Great to knock out this prerequisite to multi tabs
